client/python/conductor/FrinxConductorWrapper.py

```diff
@@ -24,140 +29,196 @@
+class TaskSource:
+    def __init__(self):
+        self.lock = threading.Lock()
+        self.task_types = {}
+
+        self.actual_uuid = None
+        self.filtered_queue = {}
+        self.actual_task_types_running = defaultdict(int)
+
+    def register_task_type(self, task_type, exec_function, domain, limit_to_thread_count):
+        self.task_types[task_type] = RegisteredWorkerTask(
+            task_type, domain, exec_function, limit_to_thread_count
+        )
+
+    def handle_tasks(self, queue):
+        with self.lock:
+            self.actual_uuid = uuid.uuid4()
+            self.filtered_queue = {
+                key: value
+                for key, value in queue.items()
+                if key in self.task_types.keys() and value > 0
+            }
+
+    def get_next_task(self, last_task_type):
+        with self.lock:
+
+            if last_task_type:
+                self.actual_task_types_running[last_task_type] -= 1
+
+            if len(self.filtered_queue) == 0:
+                return None
+
+            tasks_cannot_have_more_threads = set()
+            for task_type, task_registered in self.task_types.items():
+                task_registered: RegisteredWorkerTask
+                if (
+                    task_registered.limit_to_thread_count is not None
+                    and self.actual_task_types_running[task_type]
+                    >= task_registered.limit_to_thread_count
+                ):
+                    tasks_cannot_have_more_threads.add(task_type)
+
+            not_limited_by_thread_count = [
+                t_type
+                for t_type in self.filtered_queue.keys()
+                if t_type not in tasks_cannot_have_more_threads
+            ]
+
+            if len(not_limited_by_thread_count) == 0:
+                return None
+            task_type = random.choice(not_limited_by_thread_count)
+
+            self.actual_task_types_running[task_type] += 1
+
+            registerd_task: RegisteredWorkerTask = self.task_types[task_type]
+
+            self.filtered_queue[task_type] -= 1
+            if self.filtered_queue[task_type] <= 0:
+                self.filtered_queue.pop(task_type, None)
+
+            next_worker = NextWorkerTask()
+            next_worker.task_type = task_type
+            next_worker.exec_function = registerd_task.exec_function
+            next_worker.domain = registerd_task.domain
+            next_worker.poll_uuid = self.actual_uuid
+            return next_worker
+
+    def task_not_found_anymore(self, task_not_found):
+        if self.actual_uuid == task_not_found.poll_uuid:
+            with self.lock:
+                self.filtered_queue.pop(task_not_found.task_type, None)
+
+
+class NextWorkerTask:
+    def __init__(self):
+        self.task_type = None
+        self.domain = None
+        self.exec_function = None
+        self.poll_uuid = None
+
+
+
+class ExceptionHandlingConductorWrapper(ConductorWorker):
```

Comment:
I would keep the name FrinxConductorWrapper

since it is doing much more than ExceptionHandling now

Reply:
done